feat(): Breaking restructure

kaosat-dev opened the pull request:
WORK IN PROGRESS, DO NOT MERGE !
This is the first' break all the old stuff' PR for the V2 branch !
It 
- [x] removes most aliases
- [x] removes most cyclic dependencies
- [x] restructures a number of parts of the code to be api-centric

z3dev commented:
difference.js... really don’t like this feature...
``

z3dev commented on src/api/ops-booleans/difference.js:
```diff
@@ -0,0 +1,29 @@
+    if (isCAG(a[i])) {
+      object = object.subtract(a[i])
+    } else {
+      object = object.subtract(a[i].setColor(1, 1, 0)) // -- color the cuts
```
Can we remove this ‘feature’? It’s really annoying when doing complex designs which require colors.

kaosat-dev replied:
By feature dobyou mean just the part that colors the difference ?

z3dev commented:
Really need to decide the order of the parameters.... there’s a lot of mixing happening.

Function(objects, options)

Function(options, objects)

kaosat-dev commented:
I am systematically changing everything to (options, objects).That ordering is for a reason btw : it allows us to create curryable functions, to use the (...objects) syntax etc

z3dev commented:
function (options, objects).

I hope bought that this was the syntax required... :)

kaosat-dev commented:
Yep, always options first :-)

kaosat-dev commented on src/api/ops-booleans/union.js:
```diff
@@ -0,0 +1,45 @@
+  let i = 0
+  let a = arguments
+  if (a[0].length) a = a[0]
+  if ('extrude2d' in a[0]) { // first parameter is options
```
do you mean the implicit extrusion ? yes that will go away

z3dev replied:
Correct. No implicit anything.

kaosat-dev commented on src/api/ops-extrusions/extrusionUtils.js:
```diff
@@ -0,0 +1,265 @@
+ * @param {Number} [options.resolution=defaultResolution3D] - number of polygons per 360 degree revolution
+ * @returns {CSG} new 3D solid
+ */
+const rotateExtrude = function (cag, options) { // FIXME options should be optional
```
I am far from done with the cleanup :) but yes, only one shall remain !

z3dev commented:
I consider compactBinary()should be part of IO.

z3dev commented:
Hmmm... maybe JSON should be part of the CSG library? The difference between compact and JSON is trivial, as both are alternates to javascript object serialization. Hmmm....

z3dev commented:
Can this PR be renamed to V2API ...?

z3dev commented:
ops-booleans.js

Good, but I think we should restrict the parameters to a strict pattern, I.e. function (options, objects)

Options is REQUIRED, or detectable by testing the first argument.

Also, this check is common to all functions.

Finally, the application of the operations should be based on the first CSG/ CAG/etc. Normal procession continues without issues.

What should processing do if given MIXED object types? I think FAIL. Alternatives are... CONTINUE or BAILOUT.

kaosat-dev commented:
err , did you check out the code @z3dev  ? I already added detection and throwing an exception if you pass mixed types, and I also added the test for that :) 
Parameter orders are now changed to always be `options, objects` except for booleans where there are no options to be passed :)

kaosat-dev commented:
I understand the logic of `compactBinary` being possibly part of IO, except it is a very low level conversion, and needed by the 'core' and IO can depend on core, but not the other way around :)
for JSON : I do not know yet : the 'virtual tree' system already manipulates plain JS objects (that are almost JSON), so for the final JSON format I think we need to wait a bit for things to 'settle' :)

z3dev commented:
Colors...
- need to decide 0-255 or 0.0 to 1.0 (colors in WebGL go from 0 to 1, but possibly decided by the GL library requirements)
- hue2rgb() ?? no idea what this function does... any ideas?
- parameters to functions need to be consistent with general API

z3dev commented:
Looking at the progress...
- glad to see path2 moved into core, from math. it belongs here.
- please rename Line2.js and Line3.js to Ray2.js and Ray3.js, this will make room for real lines later.
- getTag()... just thinking but could there be problems with multiple threads calling this routine? obtaining the same number?
- toString() methods could use ES6 string templates ;)
- api, ops-transformations/mirror.js, line 23 says 'scale' ??
- core/geometry/shape*/*X.js *Y.js *Z.js ... if these are staying then move to API, core should only have matrix based functions for transforms

By the way, once you get to a fairly solid code base, with the restructure complete, then I can contribute directly with some changes / fixes. It still seems a little early. GANBATTE

z3dev commented:
Just a thought... but with the move to a functional API, we should use different verbs for get* functions. I think the use of get implies an object. Some options are...

call, conjecture, estimate, figure, gauge (also gage), guess, judge, make, place, put, reckon, suppose
conclude, deduce, extrapolate, gather, infer, reason, understand
calibrate, measure, scale
compute, work out

It should probably infer some kind of work, like compute. There are now stored values for area, center, etc.

kaosat-dev commented:
@z3dev going to merge this into the V2.0 branch despite it not being final, so you can add the changes on top ok ?
